dartagnan/src/main/java/com/dat3m/dartagnan/analysis/graphRefinement/logic/DNF.java

```diff
@@ -0,0 +1,275 @@
+    public static <V extends Literal<V>> DNF<V> FALSE() { return FALSE; }
+
+    // Using a sorted list is better (e.g. sorted by size)
+    private final Set<Conjunction<T>> cubes;
+    private int hashCode;
+
+    // This set should never be modified
+    public Set<Conjunction<T>> getCubes() {
+        return Collections.unmodifiableSet(cubes);
+    }
+
+    public int getNumberOfClauses() {
+        return cubes.size();
+    }
+
+    public int getSize() {
+        int size = 0;
+        for (Conjunction<T> cube : cubes) {
+            size += cube.getSize();
+        }
+        return size;
+    }
+
+    public boolean isFalse() {
+        return cubes.isEmpty();
+    }
+
+    // Note: This class does not reduce disjunctions of the form "p or not p" to true.
+    public boolean isTriviallyTrue() {
+        return this.equals(TRUE);
+    }
+
+    public DNF(T literal) {
+        this(new Conjunction<>(literal));
+    }
+
+    public DNF(Conjunction<T> cube) {
+        if (cube.isFalse() && FALSE != null) {
+            this.cubes = FALSE.cubes;
+            this.hashCode = FALSE.hashCode;
+        } else if (cube.isTrue() && TRUE != null) {
+            this.cubes = TRUE.cubes;
+            this.hashCode = TRUE.hashCode;
+        } else {
+            cubes = Collections.singleton(cube);
+            computeHash();
+        }
+    }
+
+    public DNF(Collection<Conjunction<T>> cubes) {
+        this.cubes = new HashSet<>(cubes);
+        reduce();
+        computeHash();
+    }
+
+    protected DNF(Set<Conjunction<T>> cubes) {
+        this(cubes, true);
+    }
+
+    protected DNF(Set<Conjunction<T>> cubes, boolean reduce) {
+        this.cubes = cubes;
+        if (reduce) {
+            reduce();
+        }
+        computeHash();
+    }
+
+    // Keeps the DNF minimal (without self-subsumption or any other advanced techniques)
+    private void reduce() {
+        for (Iterator<Conjunction<T>> it = cubes.iterator(); it.hasNext(); ) {
+            Conjunction<T> cube = it.next();
+            if (cube.isFalse()) {
+                it.remove();
+                continue;
+            } // We could also shortcut when finding a true cube
+
+            boolean isDominated = false;
+            for (Conjunction<T> cube2 : cubes) {
+                if (cube.compareToPartial(cube2) == OrderResult.GT) {
+                    isDominated = true;
+                    break;
+                }
+            }
+            if (isDominated) {
+                it.remove();
+            }
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        return hashCode;
+    }
+
+    private void computeHash() {
+        hashCode = 1;
+        for (Conjunction<T> cube : cubes) {
+            int cubeHash = cube.hashCode();
+            assert (cube.isTrue() || cubeHash != 0);
+            hashCode *= cubeHash; // Use * instead of + (important!)
+            // We assume for now that cube.hashCode() never returns 0
```

Comment:
I will move this check to ```Conjunction```. The assertion is not something that has to hold though. The code is working correctly even if 0 is returned (but it will get unperformant).